lib/bolt/application.rb

```diff
@@ -0,0 +1,612 @@
+
+      Puppet[:tasks] = false
+      ast = pal.parse_manifest(manifest_code, manifest)
+
+      if defined?(ast.body) &&
+         (ast.body.is_a?(Puppet::Pops::Model::HostClassDefinition) ||
+         ast.body.is_a?(Puppet::Pops::Model::ResourceTypeDefinition))
+        message = "Manifest only contains definitions and will result in no changes on the targets. "\
+                  "Definitions must be declared for their resources to be applied. You can read more "\
+                  "about defining and declaring classes and types in the Puppet documentation at "\
+                  "https://puppet.com/docs/puppet/latest/lang_classes.html and "\
+                  "https://puppet.com/docs/puppet/latest/lang_defined_types.html"
+        Bolt::Logger.warn("empty_manifest", message)
+      end
+
+      # Apply logging looks like plan logging
+      executor.publish_event(type: :plan_start, plan: nil)
+
+      with_benchmark do
+        pal.in_plan_compiler(executor, inventory, plugins.puppetdb_client) do |compiler|
+          compiler.call_function('apply_prep', targets)
+        end
+
+        pal.with_bolt_executor(executor, inventory, plugins.puppetdb_client) do
+          Puppet.lookup(:apply_executor).apply_ast(ast, targets, catch_errors: true, noop: noop)
+        end
+      end
+    end
+
+    # Run a command on a list of targets.
+    #
+    # @param command [String] The command.
+    # @param targets [Array[String]] The targets to run on.
+    # @param env_vars [Hash] Environment variables to set on the target.
+    # @return [Bolt::ResultSet]
+    #
+    def command_run(command, targets, env_vars: {})
+      targets = inventory.get_targets(targets)
+
+      with_benchmark do
+        executor.run_command(targets, command, env_vars: env_vars)
+      end
+    end
+
+    # Download a file from a list of targets to a directory on the controller.
+    #
+    # @param source [String] The path to the file on the targets.
+    # @param destination [String] The path to the directory on the controller.
+    # @param targets [Array[String]] The targets to run on.
+    # @return [Bolt::ResultSet]
+    #
+    def file_download(source, destination, targets)
+      destination = File.expand_path(destination, Dir.pwd)
+      targets     = inventory.get_targets(targets)
+
+      with_benchmark do
+        executor.download_file(targets, source, destination)
+      end
+    end
+
+    # Upload a file from the controller to a list of targets.
+    #
+    # @param source [String] The path to the file on the controller.
+    # @param destination [String] The destination path on the targets.
+    # @param targets [Array[String]] The targets to run on.
+    # @return [Bolt::ResultSet]
+    #
+    def file_upload(source, destination, targets)
+      future  = executor.future&.fetch('file_paths', false)
+      source  = find_file(source, future)
+      targets = inventory.get_targets(targets)
+
+      Bolt::Util.validate_file('source file', source)
```

Comment:
This needs to pass `true` as a third argument so that directories can be uploaded.
```
lucy ~/githubs/bolt ((afd2cdd0...)) $ be bolt file upload documentation/ ~/ -t localhost
The source file 'documentation/' is not a file
lucy ~/githubs/bolt ((afd2cdd0...)) $ bolt file upload documentation/ ~/ -t localhost
Started on localhost...
Finished on localhost:
  Uploaded 'documentation/' to 'localhost:/home/lucy/'
Successful on 1 target: localhost
Ran on 1 target in 0.02 sec
```

Reply:
Yep, caught this while working on getting tests working again. Good catch! :D